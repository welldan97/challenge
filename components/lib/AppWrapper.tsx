import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode, Suspense } from 'react';
import ErrorBoundary from '../../lib/ErrorBoundary';

// SECTION: Main

interface Props {
  children: ReactNode;
}

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      suspense: true,
    },
  },
});

export default function AppWrapper({ children }: Props) {
  return (
    <ErrorBoundary fallback={<h1>Error</h1>}>
      <Suspense fallback={<h1>Loading...</h1>}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </Suspense>
    </ErrorBoundary>
  );
}
