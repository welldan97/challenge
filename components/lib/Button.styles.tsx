import styled, { css } from 'styled-components';

// SECTION: Main

const styles = css`
  background: #6932d4;
  border-radius: 100px;
  color: white;
  border: none;
  height: 40px;
  width: 88px;
  appearance: none;
  text-decoration: none;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  &:hover,
  &:focus,
  &:active {
    color: white;
  }
`;

export const StyledButton = styled.button`
  ${styles}
`;
export const StyledAnchor = styled.a`
  ${styles}
`;
