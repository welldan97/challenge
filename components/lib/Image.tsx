import React, { memo } from 'react';
import BaseImage, { ImageProps } from 'next/image';

// SECTION: Main

const Image = memo(({ className, style, ...props }: ImageProps) => (
  <div className={className} style={style}>
    <BaseImage {...props} />
  </div>
));

export default Image;
