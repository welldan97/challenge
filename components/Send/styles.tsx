import styled from 'styled-components';

import Image from '../lib/Image';

// SECTION: Main

export const Success = styled(Image)`
  margin: 27px auto 11px;
  width: 100px;
`;

export const Wrapper = styled.div`
  padding-left: 24px;
  padding-right: 24px;
`;

export const Label = styled.div`
  font-size: 16px;
`;

export const Amount = styled.div`
  font-size: 28px;
  transform: translateY(-2px);
`;

export const Hr = styled.hr`
  border-top: 1px solid #202532;
  opacity: 1;
  margin-top: 12px;
  margin-bottom: 18px;
  width: 142px;
`;

export const Address = styled.div`
  transform: translateY(1px);
  margin-bottom: 14px;
`;

export const Form = styled.form`
  margin-top: 29px;
  margin-bottom: 7px;
`;
