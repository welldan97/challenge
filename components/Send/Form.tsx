import { memo } from 'react';

import Input from '../lib/Input';
import { Button, Footer, Header } from '../styles';
import { Form } from './styles';

// SECTION: Main

interface Props {
  onSend: () => void;
}

const SuccessPage = memo(({ onSend }: Props) => (
  <>
    <Header>Please fill the form to send Ethereum</Header>
    <Form onSubmit={onSend} id="transaction">
      <Input name="from" placeholder="Your address" label="From" autoFocus />
      <Input name="to" placeholder="Destination address" label="To" />
      <Input name="amount" placeholder="Ethereum amount" label="Amount" />
    </Form>
    <Footer isSeamless>
      <Button type="submit" form="transaction" data-testid="send">
        Send
      </Button>
    </Footer>
  </>
));

// /SECTION
// SECTION: Container

export default SuccessPage;
