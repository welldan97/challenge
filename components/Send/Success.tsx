import { memo } from 'react';

import { Transaction } from '../../lib/Transaction';
import { Header } from '../styles';
import { Address, Amount, Hr, Label, Success, Wrapper } from './styles';

// SECTION: Main

interface Props {
  transaction: Transaction;
}

const SuccessPage = memo(({ transaction }: Props) => (
  <>
    <Header>Transaction complete</Header>
    <Wrapper data-testid="success">
      <Success src="/Success.svg" alt="success" width="100" height="100" />
      <Label> You sent</Label>
      <Amount>{transaction.amount} ETH</Amount>
      <Hr />
      <Label>From</Label>
      <Address>{transaction.from}</Address>
      <Label>To</Label>
      <Address>{transaction.to}</Address>
    </Wrapper>
  </>
));

// /SECTION
// SECTION: Container

export default SuccessPage;
