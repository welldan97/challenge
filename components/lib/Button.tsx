import omit from 'lodash/fp/omit';
import Link from 'next/link';
import { ComponentPropsWithoutRef, memo } from 'react';
import { StyledAnchor, StyledButton } from './Button.styles';

// SECTION: Main

interface ButtonProps extends ComponentPropsWithoutRef<'button'> {
  type?: ComponentPropsWithoutRef<'button'>['type'];
}

interface AnchorProps extends ComponentPropsWithoutRef<'a'> {
  type: 'anchor';
  href: string;
}

const Button = memo((props: ButtonProps | AnchorProps = { type: 'button' }) => {
  if (props.type === 'anchor') {
    const passedHref = typeof props.href === 'string' ? props.href : '';
    return (
      <Link href={passedHref}>
        <StyledAnchor {...omit(['type'], props)} />
      </Link>
    );
  }

  return <StyledButton {...props} />;
});

export default Button;
