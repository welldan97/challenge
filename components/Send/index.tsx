import { useMutation } from '@tanstack/react-query';
import { Formik } from 'formik';
import { memo, useCallback, useMemo } from 'react';
import transactionApi from '../../api/transactionApi';
import { Transaction } from '../../lib/Transaction';
import { Logo, Box } from '../styles';
import Form from './Form';
import Success from './Success';

// SECTION: Main

interface Props {
  onSend: () => void;
  transaction?: Transaction;
}

const Send = ({ onSend, transaction }: Props) => (
  <>
    <Logo src="/Logo.svg" alt="Utrust" width="120" height="28" />
    <Box isShrunk>
      {transaction ? (
        <Success transaction={transaction} />
      ) : (
        <Form onSend={onSend} />
      )}
    </Box>
  </>
);

// /SECTION
// SECTION: Container

export default memo(() => {
  const transaction = useMemo(
    () => ({
      from: '',
      to: '',
      amount: '',
    }),
    [],
  );

  const { mutate, data } = useMutation(transactionApi.create);

  const onSubmit = useCallback(
    (nextTransaction: Transaction) => mutate(nextTransaction),
    [],
  );

  return (
    <Formik initialValues={transaction} enableReinitialize onSubmit={onSubmit}>
      {({ handleSubmit }) => <Send onSend={handleSubmit} transaction={data} />}
    </Formik>
  );
});
