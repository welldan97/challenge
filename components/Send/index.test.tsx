import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import Send from '.';
import { Response } from '../../lib/Response';
import transaction from '../../lib/test/fixtures/transaction';
import AppWrapper from '../lib/AppWrapper';
import { Transaction } from '../../lib/Transaction';

// SECTION: Main

describe('Send', () => {
  let mock: MockAdapter;

  beforeEach(() => {
    mock = new MockAdapter(axios);

    mock.onPost('/api/transactions').reply(200, {
      status: 'success',
      value: transaction,
    } as Response<Transaction>);
  });

  afterEach(() => {
    mock.reset();
  });

  test('render', () => {
    const { container } = render(
      <AppWrapper>
        <Send />
      </AppWrapper>,
    );
    expect(container).toMatchSnapshot();
  });

  test('send', async () => {
    const { container } = render(
      <AppWrapper>
        <Send />
      </AppWrapper>,
    );

    const from = screen.getByTestId('from');
    await userEvent.type(from, 'TEST_FROM');

    const to = screen.getByTestId('to');
    await userEvent.type(to, 'TEST_TO');

    const amount = screen.getByTestId('amount');
    await userEvent.type(amount, 'TEST_AMOUNT');

    const send = screen.getByTestId('send');

    await userEvent.click(send);

    await waitFor(() => {
      expect(screen.getByTestId('success')).toBeTruthy();
    });
    expect(container).toMatchSnapshot();
  });
});
