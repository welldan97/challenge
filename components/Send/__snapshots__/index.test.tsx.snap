// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send render 1`] = `
<div>
  <div
    class="sc-4d880b92-0 fwseqv"
  >
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27120%27%20height=%2728%27/%3e"
          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
        />
      </span>
      <img
        alt="Utrust"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
  </div>
  <div
    class="sc-4d880b92-1 cCwZCH"
    data-testid="wrapper"
  >
    <header
      class="sc-4d880b92-2 jLouef"
    >
      Please fill the form to send Ethereum
    </header>
    <form
      class="sc-534f3407-6 hKasLc"
      id="transaction"
    >
      <div
        class="sc-2d2a1aa8-0 cxcNIV"
      >
        <label
          class="sc-2d2a1aa8-1 haCvit"
          for="from"
        >
          From
        </label>
        <input
          class="sc-2d2a1aa8-2 eQinJl"
          data-testid="from"
          id="from"
          placeholder="Your address"
          value=""
        />
      </div>
      <div
        class="sc-2d2a1aa8-0 cxcNIV"
      >
        <label
          class="sc-2d2a1aa8-1 haCvit"
          for="to"
        >
          To
        </label>
        <input
          class="sc-2d2a1aa8-2 eQinJl"
          data-testid="to"
          id="to"
          placeholder="Destination address"
          value=""
        />
      </div>
      <div
        class="sc-2d2a1aa8-0 cxcNIV"
      >
        <label
          class="sc-2d2a1aa8-1 haCvit"
          for="amount"
        >
          Amount
        </label>
        <input
          class="sc-2d2a1aa8-2 eQinJl"
          data-testid="amount"
          id="amount"
          placeholder="Ethereum amount"
          value=""
        />
      </div>
    </form>
    <footer
      class="sc-4d880b92-3 gunIol"
    >
      <button
        class="sc-4d880b92-4 hRpFqd"
        data-testid="send"
        form="transaction"
        type="submit"
      >
        Send
      </button>
    </footer>
  </div>
</div>
`;

exports[`Send send 1`] = `
<div>
  <div
    class="sc-4d880b92-0 fwseqv"
  >
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27120%27%20height=%2728%27/%3e"
          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
        />
      </span>
      <img
        alt="Utrust"
        data-nimg="intrinsic"
        decoding="async"
        src="/Logo.svg"
        srcset="/Logo.svg 1x, /Logo.svg 2x"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
  </div>
  <div
    class="sc-4d880b92-1 cCwZCH"
    data-testid="wrapper"
  >
    <header
      class="sc-4d880b92-2 jLouef"
    >
      Transaction complete
    </header>
    <div
      class="sc-534f3407-1 cabieY"
      data-testid="success"
    >
      <div
        class="sc-534f3407-0 gksomd"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27100%27%20height=%27100%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="success"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
      <div
        class="sc-534f3407-2 gTneqL"
      >
         You sent
      </div>
      <div
        class="sc-534f3407-3 feBzYV"
      >
        TEST_AMOUNT
         ETH
      </div>
      <hr
        class="sc-534f3407-4 gWNUMj"
      />
      <div
        class="sc-534f3407-2 gTneqL"
      >
        From
      </div>
      <div
        class="sc-534f3407-5 hezqVB"
      >
        TEST_FROM
      </div>
      <div
        class="sc-534f3407-2 gTneqL"
      >
        To
      </div>
      <div
        class="sc-534f3407-5 hezqVB"
      >
        TEST_TO
      </div>
    </div>
  </div>
</div>
`;
